### PR TITLE
Add a SearchProvider test for matching an empty string

### PR DESCRIPTION
**Pull request by kevinhinterlong**
This was brought up in #390 and verifies that searching for an empty string will match all results

### CODE REVIEW

**Review by michael-mclawhorn — APPROVED** [review]
Looks good.

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy [hunk 1]
```diff
@@ -526,6 +526,31 @@ abstract class SearchProviderSpec<T extends SearchProvider> extends Specificatio
...
+        setup: "Given a filter that will match every row"
+        /* Expected rows, not necessarily in this order:
+                name: "owl", description: "this is an owl"
+                name: "hawk", description: "this is a raptor"
+                name: "eagle", description: "this is a raptor"
+                name: "kumquat", description: "this is not an animal"
+        */
+        Set<ApiFilter> filters = [
+                buildFilter("animal|desc-startswith[this]"),
+                buildFilter("animal|desc-contains[\"\"]")
```
Comment (QubitPi) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy]: Avoid escaping quotes - `'animal|desc-contains[""]'`

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy [hunk 2]
```diff
@@ -526,6 +526,31 @@ abstract class SearchProviderSpec<T extends SearchProvider> extends Specificatio
...
+        ]
+        and: "We get the second page, where each page has two rows (so both pages should have two results)"
```
Comment (QubitPi) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy]: `We get the second page, where each page has two rows` sounds good enough already

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy [hunk 3]
```diff
@@ -526,6 +526,31 @@ abstract class SearchProviderSpec<T extends SearchProvider> extends Specificatio
...
+        PaginationParameters parameters = new PaginationParameters(2, 2)
+
+        when: "We query the search provider"
+        Pagination<DimensionRow> resultsPage = searchProvider.findFilteredDimensionRowsPaged(filters, parameters)
+
+        then: "We get only the last page of results (which has two)"
```
Comment (QubitPi) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy]: `We get only the last page of results` sounds good enough already

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by QubitPi** [comment]
Looks pretty good. Minor comments

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy [hunk 4]
```diff
@@ -526,6 +526,31 @@ abstract class SearchProviderSpec<T extends SearchProvider> extends Specificatio
...
     }
 
+    def "An empty string filter value should return all results"() {
+        setup: "Given a filter that will match every row"
+        /* Expected rows, not necessarily in this order:
+                name: "owl", description: "this is an owl"
+                name: "hawk", description: "this is a raptor"
+                name: "eagle", description: "this is a raptor"
+                name: "kumquat", description: "this is not an animal"
+        */
```
Comment (QubitPi) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy]: It's good to list expected rows. It's gonna be even better if you could code them into test, i.e. making sure the 4 rows are equal to those

**Review by ghost — CHANGES_REQUESTED** [review]

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy (hunk 4 above)
Comment (kevinhinterlong) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy]: I don't think there's any guarantee  about order they're returned in. Do you want me to do something like
```
Set<String> expectedOutcomes = ["owl", "hawk", "eagle", "kumquat"]
...

Set<String> actualOutcomes = resultsPage.pageOfData.stream().map { it.getKeyValue() }.collect(Collectors.toSet())
actualOutcomes.equals(expectedOutcomes)
```

Comment (QubitPi) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy]: You could try
```
Collections.sort(actualOutcomes);
Collections.sort(expectedOutcomes);
```

Comment (kevinhinterlong) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy]: So you want me to make them both lists, sort them like
```java
List<String> actualOutcomes = resultsPage.pageOfData.stream().map { it.getKeyValue() }.sorted().collect(Collectors.toList())
Collections.sort(expectedOutcomes);
```
then check for equality?
I prefer the Set way because there's no need to rely on sorting the results

Comment (QubitPi) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy]: You can do that

Comment (michael-mclawhorn) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy]: Ideally we'd like the code going through the system to preserve or coerce ordering.  Without that we'll be inefficient in terms of creating cache keys.  
I'd suggest a jira ticket to create a Comparator on ApiFilter, probably sorting on DimensionName, DimensionFieldName, OperationName, and then lexically sorted values.  Then the ApiFilters could be stored in a SortedSet, which should make test cases far easier to write.

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy [hunk 5]
```diff
@@ -526,6 +529,34 @@ abstract class SearchProviderSpec<T extends SearchProvider> extends Specificatio
...
+        Pagination<DimensionRow> resultsPage = searchProvider.findFilteredDimensionRowsPaged(filters, parameters)
+
+        then: "We get only one page of results"
+        resultsPage.getPageOfData().size() == 4
+
+        and: "We get only {'owl', 'hawk', 'eagle', 'kumquat'}"
+        resultsPage.numResults == 4
+        Set<String> actualOutcomes = resultsPage.pageOfData.stream().map { it.getKeyValue() }.collect(Collectors.toSet())
+        actualOutcomes.equals(expectedOutcomes)
+    }
```
Comment (michael-mclawhorn) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy]: To make this test more idiomatic for groovy and spock, lines 564-5 could be rewritten as follows:
expectedOutcomes == resultsPage.pageOfData.collect { it.getKeyValue() } as Set

This uses the simpler groovy collect without going through java streams of collectors library and the coercion operator to control the collection type.
The use of '==' is (sort of) equivalent to .equals but leverages the power assert capability for more useful debugging info.

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy (hunk 4 above)
Comment (michael-mclawhorn) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy]: I'm fine with set comparison here.  More on how to streamline the test below.

**Review by michael-mclawhorn — CHANGES_REQUESTED** [review]
I'd love to see the groovy-ification of the tests.

**Review by michael-mclawhorn — APPROVED** [review]

**Review by archolewa — APPROVED** [review]

**Comment by michael-mclawhorn** [comment]
@QubitPi  please update your feedback

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy [hunk 6]
```diff
@@ -25,6 +25,9 @@ import org.joda.time.DateTime
 
 import spock.lang.Specification
 
+
```
Comment (QubitPi) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/impl/SearchProviderSpec.groovy]: delete new line

**Review by ghost — APPROVED** [review]
